AS-513: Run analyses on snapshot tables

MatthewBemis opened the pull request:
Tested these scenarios:

Switching back and forth between snapshot and table sources to make sure everything looked correct (did not see issues)
Reloading the page and ensuring that snapshot dropdowns are properly loaded (dropdowns re-populated correctly based on existing MC)
Saving inputs/outputs that are invalid for that snapshot (showed validation errors, as expected)
Running an analysis that generated 4 workflows (completed successfully)
Trying to save and run an analysis on a snapshot without a table selected (buttons grayed out)

MatthewBemis committed:
get second dropdown going

MatthewBemis committed:
stash

MatthewBemis committed:
mostly working, run button does't work yet

MatthewBemis committed:
cleanup

MatthewBemis committed:
so close

MatthewBemis committed:
get launching working

MatthewBemis committed:
bit of cleanup

MatthewBemis committed:
bit of cleanup

MatthewBemis committed:
update comment

MatthewBemis committed:
formatting

MatthewBemis commented on src/libs/analysis.js:
```diff
@@ -61,3 +61,21 @@ export const launch = async ({
   })
 }
+
+export const launchSnapshot = async ({
```
I swing back and forth on wanting to honor DRY and not wanting to have conditionals _literally_ all over the place. I find that in the WorkflowView component, some of the logic may have gotten to be a bit convoluted or complicated because of the conditionals.

MatthewBemis commented on src/pages/workspaces/workspace/workflows/LaunchAnalysisModal.js:
```diff
@@ -80,7 +88,8 @@ const LaunchAnalysisModal = ({
     [type === chooseSetComponents, () => 1],
-    [type === processMergedSet, () => _.flow(mergeSets, _.uniqBy('entityName'))(selectedEntities).length]
+    [type === processMergedSet, () => _.flow(mergeSets, _.uniqBy('entityName'))(selectedEntities).length],
+    [type === processSnapshotTable, () => 1] //TODO: could be table rowCount but rowCount to workflow conversion is currently unclear
```
Using `rowCount` from the snapshot table details is probably the right move in the end, but these submissions were spawning 0 workflows (but still writing outputs). I've followed the examples in the Rawls unit tests but I think there may be something missing either here or there.

MatthewBemis commented on src/pages/workspaces/workspace/workflows/WorkflowView.js:
```diff
@@ -482,10 +526,13 @@ const WorkflowView = _.flow(
         currentSnapRedacted: isRedacted, savedSnapRedacted: isRedacted,
         entityMetadata,
+        availableSnapshots: snapshots,
+        selectedEntitySource,
```
Would a more appropriate piece of state be to make this a boolean called `snapshotSourceSelected` rather than doing string comps everywhere?

zarsky-broad committed:
some simplification

Co-Authored-By: Matt Bemis <MatthewBemis@users.noreply.github.com>

MatthewBemis committed:
improved launch reason logic

MatthewBemis committed:
handle populating existing dropdown values on load

zarsky-broad reviewed (commented):
One thing before I review completely:
When we're setting the selection to a snapshot, we should probably set `selectedSnapshotTableNames: undefined, selectedTableName: undefined`

davidangb commented:
per the PR template, please indicate what testing you've already done!

MatthewBemis commented:
@davidangb thanks, I just updated the description!

davidangb commented on src/pages/workspaces/workspace/workflows/WorkflowView.js:
```diff
@@ -705,33 +711,74 @@ const WorkflowView = _.flow(
+                    this.setState(_.unset(['modifiedConfig', 'entityName']))
+                  },
+                  styles: { container: old => ({ ...old, display: 'inline-block', width: 200, marginLeft: '0.5rem' }) },
+                  options: selectedSnapshotTableNames
```
could you make this dropdown sorted alphabetically? Otherwise it's very difficult for a user to find the table they're  searching for. `options: _.sortBy(_.identity)(selectedSnapshotTableNames)` should do it, but UI experts might have a better way and/or you might want to sort the array earlier in the code flow.

davidangb commented on src/pages/workspaces/workspace/workflows/WorkflowView.js:
```diff
@@ -705,33 +711,74 @@ const WorkflowView = _.flow(
+                  onChange: async ({ value, source }) => {
+                    if (source === 'snapshot') {
+                      const snapshotId = _.find({ name: value }, availableSnapshots).reference.snapshot
+                      const { tables } = await Ajax(signal).DataRepo.snapshot(snapshotId).details()
```
1. could you cache the result of this somewhere, so we don't query for the same data every time if a user swaps the root entity type selection?
2. consider calling Rawls' entity type metadata endpoint, specifying the snapshot reference name, instead of calling data repo directly. It's the same data, but Rawls is already parsing it out for you. And if in the future Rawls supports entity types that are not 1:1 with table names - for instance, if Rawls supports an entity type that is the result of a join - you'd automatically inherit that here by calling Rawls instead of DR.

davidangb replied:
further - if you called Rawls' entity type metadata endpoint instead, you'd also get all the attribute names for each entity type, which would allow you to satisfy the typeahead&validation when editing the method config inputs/outputs

davidangb commented on src/pages/workspaces/workspace/workflows/WorkflowView.js:
```diff
@@ -705,33 +711,74 @@ const WorkflowView = _.flow(
+                  id,
+                  isDisabled: !!Utils.editWorkspaceError(ws),
+                  isClearable: false,
+                  isSearchable: false,
```
why disable `isSearchable`? Would be nice to be able to type to find the entity type I'm looking for.

davidangb reviewed (commented):
initial comments inline!

MatthewBemis committed:
allow for searching, sort options

MatthewBemis committed:
handle undefineds and fix bug with save button

MatthewBemis committed:
typo

MatthewBemis committed:
cache values

MatthewBemis committed:
populate typeahead

MatthewBemis committed:
can no longer make certain assumptions about _id attr names

MatthewBemis committed:
remove console.log

MatthewBemis commented:
@davidangb @zarsky-broad this is ready for review

i tested that once we loaded snapshot metadata once, it was cached and not loaded again. i also tested that input suggestions work for snapshots

MatthewBemis committed:
fix bugs with typeahead

MatthewBemis committed:
configure default state better

davidangb reviewed (approved):
works for me! does the last remaining "Potentially inconsistent state update" warning need addressing?

MatthewBemis committed:
fix LGTM warning

davidangb reviewed (approved):
still :+1:

panentheos commented on src/pages/workspaces/workspace/workflows/WorkflowView.js:
```diff
@@ -478,14 +470,27 @@ const WorkflowView = _.flow(
+        availableSnapshots: snapshots,
+        selectedEntitySource: modifiedConfig.dataReferenceName ? 'snapshot' : 'table',
+        selectedTableName: modifiedConfig.dataReferenceName ? modifiedConfig.rootEntityType : undefined,
+        cachedSnapshotEntityMetadata: modifiedConfig.dataReferenceName ? [{ snapshotName: modifiedConfig.dataReferenceName, metadata: selectedSnapshotEntityMetadata }] : [{}],
```
Initializing this to `[{}]` adds an empty object to the cache, which will never match but might be confusing when debugging. Just an empty array `[]` should be fine.

panentheos commented on src/pages/workspaces/workspace/workflows/WorkflowView.js:
```diff
@@ -478,14 +470,27 @@ const WorkflowView = _.flow(
+        selectedTableName: modifiedConfig.dataReferenceName ? modifiedConfig.rootEntityType : undefined,
+        cachedSnapshotEntityMetadata: modifiedConfig.dataReferenceName ? [{ snapshotName: modifiedConfig.dataReferenceName, metadata: selectedSnapshotEntityMetadata }] : [{}],
+        selectedSnapshotTableNames,
+        selectedSnapshotEntityMetadata,
```
Since these two pieces of information can be derived from state we already have, it would be better to compute them dynamically as needed, instead of persisting them to state, since adding extra state opens up more ways things might accidentally get out of sync, either now or in the future.

panentheos commented on src/pages/workspaces/workspace/workflows/WorkflowView.js:
```diff
@@ -478,14 +470,27 @@ const WorkflowView = _.flow(
         entityMetadata,
+        availableSnapshots: snapshots,
+        selectedEntitySource: modifiedConfig.dataReferenceName ? 'snapshot' : 'table',
+        selectedTableName: modifiedConfig.dataReferenceName ? modifiedConfig.rootEntityType : undefined,
```
We can compute these two things dynamically from the `modifiedConfig`, so it would be better to avoid storing them as extra state.

panentheos commented on src/pages/workspaces/workspace/workflows/WorkflowView.js:
```diff
@@ -557,13 +562,14 @@ const WorkflowView = _.flow(
-    const { modifiedConfig: { rootEntityType } } = this.state
-    return this.isSingle() || !!rootEntityType
+    const { modifiedConfig: { rootEntityType, dataReferenceName } } = this.state
+    return (this.isSingle() || !!rootEntityType) || (!!dataReferenceName && !!rootEntityType)
```
Is this logic right? It looks like `!!rootEntityType` will always cause the whole expression to be true, regardless of `!!dataReferenceName`

MatthewBemis replied:
I believe that it's right. If you're part way through configuring the analysis to run on a snapshot (i.e. you've selected a snapshot name but not a table name) the expression will be `false` because `!!dataReferenceName` = `true`, `!!rootEntityType` = `false`, and `this.isSingle()` will evaluate to `false`

There was a bug that I fixed in a recent commit where the overall expression was evaluating to `true` when it shouldn't have because `rootEntityType` was never unset when selecting a snapshot name, so if you were testing locally that may have been the cause

panentheos commented on src/pages/workspaces/workspace/workflows/WorkflowView.js:
```diff
@@ -478,14 +470,27 @@ const WorkflowView = _.flow(
         errors: isRedacted ? { inputs: {}, outputs: {} } : augmentErrors(validationResponse),
-        entitySelectionModel: this.resetSelectionModel(modifiedConfig.rootEntityType, readSelection ? selection.entities : {}, entityMetadata),
+        entitySelectionModel: this.resetSelectionModel(
+          modifiedConfig.dataReferenceName ? modifiedConfig.dataReferenceName : modifiedConfig.rootEntityType,
```
Is this right? Though I think this value ends up getting ignored in snapshot mode, I would expect this to be `rootEntityType` in that case. Presumably this will become relevant when we add the ability to select snapshot table entities.

panentheos commented on src/pages/workspaces/workspace/workflows/WorkflowView.js:
```diff
@@ -478,14 +470,27 @@ const WorkflowView = _.flow(
+        entitySelectionModel: this.resetSelectionModel(
+          modifiedConfig.dataReferenceName ? modifiedConfig.dataReferenceName : modifiedConfig.rootEntityType,
+          readSelection ? selection.entities : {},
+          entityMetadata, modifiedConfig.dataReferenceName ? 'snapshot' : ''
```
If the last argument is a 'source', we should probably pass `'table'` if we're in table mode, for consistency.

panentheos commented on src/pages/workspaces/workspace/workflows/WorkflowView.js:
```diff
@@ -705,33 +713,76 @@ const WorkflowView = _.flow(
+                      this.setState({
+                        selectedSnapshotEntityMetadata, selectedEntityType: value, selectedTableName: undefined, selectedSnapshotTableNames: _.keys(selectedSnapshotEntityMetadata),
+                        entitySelectionModel: this.resetSelectionModel(value, {}, {}, source), selectedEntitySource: source,
+                        cachedSnapshotEntityMetadata: _.concat(cachedSnapshotEntityMetadata, { snapshotName: value, metadata: selectedSnapshotEntityMetadata })
```
This will add unconditionally add a new cache entry every time the selection changes, causing it to grow indefinitely. This could be confusing to debug, and would cause memory leaks in certain usage patterns.

panentheos commented on src/pages/workspaces/workspace/workflows/WorkflowView.js:
```diff
@@ -705,33 +713,76 @@ const WorkflowView = _.flow(
-                    this.setState({ entitySelectionModel: this.resetSelectionModel(value) })
+                  onChange: async ({ value, source }) => {
+                    if (source === 'snapshot') {
+                      const selectedSnapshotCachedMetadata = _.find({ snapshotName: value }, cachedSnapshotEntityMetadata)
```
TOL: If we're just doing plain caching here, it might be simpler to use something like `memoizeAsync` instead of managing the cache by hand.

panentheos commented on src/pages/workspaces/workspace/workflows/WorkflowView.js:
```diff
@@ -705,33 +713,76 @@ const WorkflowView = _.flow(
-                    onClick: () => this.setState({ selectingData: true })
-                  }, ['Select Data']),
-                  label({ style: { marginLeft: '1rem' } }, [`${this.describeSelectionModel()}`])
+              (entitySelectionModel.type === processSnapshotTable) ? h(IdContainer, [id => div({ style: { margin: '2rem 0 0 2rem' } }, [
```
`IdContainer` is used to generate a unique id to associate a label. Since there's no label here, the id isn't doing anything and could be removed.

Also TOL: It doesn't look like the new select has any descriptive text (label or aria-label), so we could probably do better from an accessibility perspective.

panentheos commented on src/pages/workspaces/workspace/workflows/WorkflowView.js:
```diff
@@ -705,33 +713,76 @@ const WorkflowView = _.flow(
+                    this.setState(_.unset(['modifiedConfig', 'entityName']))
+                  },
+                  styles: { container: old => ({ ...old, display: 'inline-block', width: 200, marginLeft: '0.5rem' }) },
+                  options: _.sortBy(_.identity)(selectedSnapshotTableNames)
```
nit: if you're passing multiple params at once to a curried lodash function, you can just make a single call, e.g. `_.sortBy(_.identity, selectedSnapshotTableNames)`

MatthewBemis committed:
lower-hanging fruit first, from PR review

MatthewBemis committed:
fix a bug, remove IdContainer, add aria label

MatthewBemis committed:
prevent cache from growing infinitely

MatthewBemis committed:
fix bug with save button

zarsky-broad commented on src/pages/workspaces/workspace/workflows/WorkflowView.js:
```diff
@@ -375,21 +376,10 @@ const WorkflowView = _.flow(
     this.setState({
       processSingle: !config.rootEntityType,
-      selectedEntityType: config.rootEntityType
+      selectedEntityType: config.dataReferenceName ? config.dataReferenceName : config.rootEntityType
```
This could actually be simplified to:
```suggestion
      selectedEntityType: config.dataReferenceName || config.rootEntityType
```

zarsky-broad commented on src/pages/workspaces/workspace/workflows/WorkflowView.js:
```diff
@@ -478,14 +470,27 @@ const WorkflowView = _.flow(
         errors: isRedacted ? { inputs: {}, outputs: {} } : augmentErrors(validationResponse),
-        entitySelectionModel: this.resetSelectionModel(modifiedConfig.rootEntityType, readSelection ? selection.entities : {}, entityMetadata),
+        entitySelectionModel: this.resetSelectionModel(
+          modifiedConfig.dataReferenceName ? modifiedConfig.dataReferenceName : modifiedConfig.rootEntityType,
```
```suggestion
          modifiedConfig.dataReferenceName || modifiedConfig.rootEntityType,
```

zarsky-broad commented on src/pages/workspaces/workspace/workflows/WorkflowView.js:
```diff
@@ -478,14 +470,27 @@ const WorkflowView = _.flow(
+        availableSnapshots: snapshots,
+        selectedEntitySource: modifiedConfig.dataReferenceName ? 'snapshot' : 'table',
+        selectedTableName: modifiedConfig.dataReferenceName ? modifiedConfig.rootEntityType : undefined,
+        cachedSnapshotEntityMetadata: modifiedConfig.dataReferenceName ? [{ snapshotName: modifiedConfig.dataReferenceName, metadata: selectedSnapshotEntityMetadata }] : [],
```
Why not simplify the structure of this state to this:
```suggestion
        cachedSnapshotEntityMetadata: modifiedConfig.dataReferenceName ? { [modifiedConfig.dataReferenceName]: selectedSnapshotEntityMetadata } : {},
```

Then, you won't have to `_.find` to check if something is there, just see if there's a value for the key.

zarsky-broad reviewed (changes requested):
Strong +1 to Brett's comments about avoiding computed state and replacing caching with `memoizeAsync`.

MatthewBemis committed:
Update src/pages/workspaces/workspace/workflows/WorkflowView.js

Co-authored-by: Isaac Zarsky <zarsky@broadinstitute.org>

MatthewBemis committed:
Update src/pages/workspaces/workspace/workflows/WorkflowView.js

Co-authored-by: Isaac Zarsky <zarsky@broadinstitute.org>

MatthewBemis committed:
remove selectedEntitySource from state

MatthewBemis committed:
remove selectedTableName from state

MatthewBemis committed:
basic memoize, still need to strip out prior cache

MatthewBemis committed:
use memoizeAsync

MatthewBemis commented:
@zarsky-broad please take another look. I believe I've addressed all comments left by you and Brett at this point.

It seems to be working correctly but please pay careful attention to my use of memoizeAsync for snapshotEntityMetadata :)

MatthewBemis committed:
remove selectedSnapshotTableNames from state

MatthewBemis committed:
pr feedback w/ isaac

MatthewBemis committed:
logic got reversed in IO table

MatthewBemis committed:
use row count to show how many analyses will be launched

MatthewBemis committed:
simplify new logic and unset selectedSnapshotEntityMetadata appropriately
